ocaml/utils/misc.ml

```diff
@@ -215,6 +215,29 @@ module Stdlib = struct
           }
       in
       find_prefix ~longest_common_prefix_rev:[] first second
+
+    let [@inline] merge_iter ~cmp ~left_only ~right_only ~both t1 t2 =
+      let rec loop t1 t2 =
+        match t1, t2 with
+        | [], [] -> ()
+        | a :: t1', [] -> left_only a; loop t1' []
+        | [], b :: t2' -> right_only b; loop [] t2'
+        | a :: t1', b :: t2' ->
+            match cmp a b with
+            | 0 -> both a b; loop t1' t2'
+            | c when c < 0 -> left_only a; loop t1' t2
+            | _ -> right_only b; loop t1 t2'
+      in
+      loop t1 t2
+
+    let [@inline] merge_map ~cmp ~left_only ~right_only ~both t1 t2 =
+      let acc_rev = ref [] in
```

Comment:
They don't just allow the lists to be of different lengths. They also match up elements that compare as equal. If the lists are sorted lists of key-value pairs and `cmp` just compares the keys, then this is basically `Map.merge`.

I'll add comments to the .mli to make all this clearer. (Also, it might be easier just to have the function explicitly work on key-value pairs; we can always generalize the type again if we ever need to.)

Reply:
Oh, but you're still right that this would be a bit nicer as a fold and then I don't have to mess with a ref to specialize it. Coding that up now.